components/raftstore/src/store/peer.rs

```diff
@@ -461,6 +461,39 @@ pub enum ForceLeaderState {
     ForceLeader { failed_stores: HashSet<u64> },
 }
 
+// Following shared states are used while reporting to PD for unsafe recovery and shared among
+// all the regions per their life cycle.
+// The work flow is like:
+//                   start_unsafe_recovery_report
+//                                 |
+//            --------------------------------------------
+//           |                  |                         |
+//      wait apply          wait apply       ...      wait apply
+//           |                  |                         |
+//            --------------------------------------------
+//                                 |
+//            --------------------------------------------
+//           |                  |                         |
+//     fill out report    fill out report    ...    fill out report
+//           |                  |                         |
+//            --------------------------------------------
+//                                 |
+//                     send report (store heartbeat)
+#[derive(Copy, Clone)]
+pub struct UnsafeRecoveryReportId {
```

Comment:
You can write like this to reduce the verbose logic elsewhere. Other places only need to do `shared_state.clone()`, and `shared_state.finish(router)`.


```rust
pub struct UnsafeRecoveryContext {
    pub step_id: u64,
    pub exit_force_leader: bool,
    pub counter: AtomicUsize,
}

pub struct UnsafeRecoveryPlanExecutionSharedState(Arc<Mutex<UnsafeRecoveryContext>>);

impl Clone for UnsafeRecoveryPlanExecutionSharedState {
    fn clone(&self) -> UnsafeRecoveryPlanExecutionSharedState {
        self.0.lock().unwrap().counter.fetch_add(1, Ordering::SeqCst);
        UnsafeRecoveryPlanExecutionSharedState(self.0.clone())
    }
}

impl UnsafeRecoveryPlanExecutionSharedState {
    fn finish(&mut self, router: &RaftRouter) {
        if self.counter.fetch_sub(1, Ordering::SeqCst) == 1 {
            wait_apply_for_unsafe_recovery_report(
                router,
                &self.ctx,
            }
        }
    }
}

pub struct UnsafeRecoveryWaitApplySharedState(Arc<Mutex<UnsafeRecoveryContext>>);

impl Clone for UnsafeRecoveryWaitApplySharedState {
    fn clone(&self) -> UnsafeRecoveryWaitApplySharedState {
        self.0.lock().unwrap().counter.fetch_add(1, Ordering::SeqCst);
        UnsafeRecoveryWaitApplySharedState(self.0.clone())
    }
}

impl UnsafeRecoveryWaitApplySharedState {
    fn finish(&mut self, router: &RaftRouter) {
        if self.0.lock().unwrap().counter.fetch_sub(1, Ordering::SeqCst) == 1 {
            fill_out_unsafe_recovery_report(
                router,
                &self.0.lock().unwrap().ctx,
            }
        }
    }
}

pub struct UnsafeRecoveryCollectReportSharedState(Arc<Mutex<UnsafeRecoveryCollectReportSharedStateInner>>);

struct UnsafeRecoveryCollectReportSharedStateInner {
    reports: Vec<pdpb::PeerReport>,
    ctx: UnsafeRecoveryContext,
}

impl Clone for UnsafeRecoveryWaitApplySharedState {
    fn clone(&self) -> UnsafeRecoveryWaitApplySharedState {
        self.0.lock().unwrap().ctx.counter.fetch_add(1, Ordering::SeqCst);
        UnsafeRecoveryWaitApplySharedState(self.0.clone())
    }
}

impl UnsafeRecoveryCollectReportSharedState {
    fn finish(&mut self, router: &RaftRouter) {
        let inner = self.0.lock().unwrap();
        if inner.counter.fetch_sub(1, Ordering::SeqCst) == 1 {
            schedule_store_report(
                router,
                &inner.ctx,
            }
        }
    }
}
```